util/linuxfw: add package to gather Linux firewall information

andrew-d opened the pull request:
This package is an initial implementation of something that can read
netfilter and iptables rules from the Linux kernel without needing to
shell out to an external utility; it speaks directly to the kernel using
syscalls and parses the data returned.

Currently this is read-only since it only knows how to parse a subset of
the available data.

Signed-off-by: Andrew Dunham <andrew@tailscale.com>

maisem reviewed (commented):
I don't know if we should do this via Hostinfo, maybe c2n?

bradfitz commented on cmd/tailscale/depaware.txt:
```diff
@@ -7,6 +7,12 @@ tailscale.com/cmd/tailscale dependencies: (generated by github.com/tailscale/dep
    W 💣 github.com/alexbrainman/sspi/negotiate                       from tailscale.com/net/tshttpproxy
         github.com/fxamacker/cbor/v2                                 from tailscale.com/tka
         github.com/golang/groupcache/lru                             from tailscale.com/net/dnscache
+   L    github.com/google/nftables                                   from tailscale.com/util/linuxfw
```
the CLI binary bringing in all this is a little surprising. And gvisor stuff below.

bradfitz commented on hostinfo/hostinfo_linux.go:
```diff
@@ -168,3 +175,72 @@ func packageTypeLinux() string {
+		BinInfo: make(map[string]tailcfg.LinuxFWBinInfo),
+	}
+
+	n, err := linuxfw.DetectIptables()
```
why don't we put all this detection in its own package outside of hostinfo and have it register itself with the hostinfo package in its init?

then only tailscaled can empty-import it for its side effects.

bradfitz commented:
> I don't know if we should do this via Hostinfo, maybe c2n?

We're going to want to know this early at runtime for our own purposes in the Linux router code to make decisions on how to program iptables from containers.

@andrew-d, these commits/PR should reference #5621

bradfitz commented on tailcfg/tailcfg.go:
```diff
@@ -721,6 +723,45 @@ func appendStructPtrDiff(base []string, pfx string, p1, p2 reflect.Value) (ret [
+	// determine its version.
+	Error string `json:",omitempty"`
+	// Flavor is whether we were able to detect a "flavor" of the binary;
+	// this is only set for the "iptables" and "ip6tables" binaries.
```
and when it's set, what are the set/form of values it can be?

maisem commented:
> We're going to want to know this early at runtime for our own purposes in the Linux router code to make decisions on how to program iptables from containers.

Sure, but is hostinfo really the place to put it? This should just be on the link monitor and be refreshed on link changes

andrew-d commented:
@bradfitz @maisem - I rebased and updated this to only add the `util/linuxfw` package, not plumbing it into the rest of `tailscaled`, so that a future merge is smaller and more self-contained. PTAL?

bradfitz commented on util/linuxfw/linuxfw.go:
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package linuxfw returns the kind of firewall being used by the kernel
```
missing trailing period

bradfitz commented on util/linuxfw/helpers.go:
```diff
@@ -0,0 +1,44 @@
+	"unicode"
+)
+
+func btos(b []byte) string {
```
https://pkg.go.dev/golang.org/x/sys/unix#ByteSliceToString instead?

bradfitz commented on util/linuxfw/iptables.go:
```diff
@@ -0,0 +1,821 @@
+
+	"github.com/josharian/native"
+	"golang.org/x/sys/unix"
+	linux_abi "gvisor.dev/gvisor/pkg/abi/linux"
```
underscore is a bit jarring/atypical

bradfitz commented on util/linuxfw/iptables.go:
```diff
@@ -0,0 +1,821 @@
+	return nil
+}
+
+// DetectIptables returns the number of iptables rules that are present in the
+// system, ignoring the default "ACCEPT" rule present in the standard iptables
+// chains.
+func DetectIptables() (int, error) {
```
document: when does it return an error?

on 0? on error? on not supported by kernel?

bradfitz commented on util/linuxfw/iptables.go:
```diff
@@ -0,0 +1,821 @@
+	ret.NFCache = native.Endian.Uint32(p.getBytes(4))
+	targetOffset := int(native.Endian.Uint16(p.getBytes(2)))
+	nextOffset := int(native.Endian.Uint16(p.getBytes(2)))
+	/* Comeback = */ p.getBytes(4)
```
comeback?

andrew-d replied:
Unused field; that's the name that the kernel calls it. Expanded the comment a bit.

bradfitz commented on util/linuxfw/iptables.go:
```diff
@@ -0,0 +1,821 @@
+	}
+
+	if ip.InputInterface != "" {
+		fmt.Fprintf(&sb, `, InputInterface:%s/%s`, ip.InputInterface, translateMask(ip.InputInterfaceMask))
```
not a fan of backticks for string literals except 1) in regexps always, even when not needed, 2) when really needed because of newlines or lots of double quotes

andrew-d replied:
Switched everything to regular double-quotes.

bradfitz commented on util/linuxfw/linuxfw_unsupported.go:
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !linux || (linux && (386 || loong64))
```
the `linux &&` part is redundant but fine I guess
